external-crates/move/crates/move-compiler/src/typing/deprecation_warnings.rs

```diff
@@ -0,0 +1,564 @@
+            locs.reverse();
+
+            let initial_diag = locs.pop().expect("ICE: locs should not be empty");
+
+            let module_msg = if deprecation.location == AttributePosition::Module {
+                ". It is deprecated since its whole module is marked deprecated"
+            } else {
+                ""
+            };
+
+            let location_string = match deprecation.location {
+                AttributePosition::Module => {
+                    format!("This {} is deprecated{}", initial_diag.0, module_msg,)
+                }
+                x => format!("This {} is deprecated", x),
+            };
+
+            let initial_message = match deprecation.deprecation_note {
+                None => location_string,
+                Some(note) => format!("{location_string}: {note}"),
+            };
+
+            let mut diag = diag!(
+                TypeSafety::DeprecatedUsage,
+                (initial_diag.1, initial_message)
+            );
+
+            for (position, loc) in locs {
+                diag.add_secondary_label((loc, format!("Deprecated {position} also used here")));
+            }
+            self.env.add_diag(diag);
+        }
+    }
+}
+
+// If `ty` is a qualified datatype name, return the module ident, datatype name, and location.
+// Otherwise, return None.
+fn qualified_datatype_name_of_type(ty: &Type) -> Option<(&ModuleIdent, &DatatypeName)> {
+    match &ty.value {
+        Type_::Apply(_, sp!(_, N::TypeName_::ModuleType(mident, name)), _) => Some((mident, name)),
+        _ => None,
+    }
+}
+
+// Process the deprecation attributes for a given member (module, constant, function, etc.) and
+// return `Optiong<Deprecation>` if there is a #[deprecated] attribute. If there are invalid
+// #[deprecated] attributes (malformed, or multiple on the member), add an error diagnostic to
+// `env` and return None.
+fn deprecations(
+    env: &mut CompilationEnv,
+    attr_position: AttributePosition,
+    attrs: &E::Attributes,
+    source_location: Loc,
+    mident: ModuleIdent,
+) -> Option<Deprecation> {
+    let deprecations: Vec<_> = attrs
+        .iter()
+        .filter(|(_, v, _)| matches!(v, KnownAttribute::Deprecation(_)))
+        .collect();
+
+    if deprecations.is_empty() {
+        return None;
+    }
+
+    assert!(
+        deprecations.len() == 1,
+        "ICE: verified that there is at at least one deprecation attribute above, \
+        and expansion should have failed if there were multiple deprecation attributes."
+    );
+
+    let (loc, _, attr) = deprecations
+        .last()
+        .expect("Verified deprecations is not empty above");
+
+    let mut make_invalid_deprecation_diag = || {
+        let mut diag = diag!(
+            Attributes::InvalidUsage,
+            (*loc, "Invalid 'deprecated' attribute")
+        );
+        let note = "Deprecation attributes must be written as `#[deprecated]` or `#[deprecated(note = b\"message\")]`";
```

Comment:
Consider using `DeprecationAttribute::DEPRECATED` for these to avoid name drift.